### PR TITLE
Wrap CP-SAT assumptions

### PR DESCRIPTION
**Pull request by stradivari96**
Some questions:
- Is this the desired API or maybe `model.AddAssumptions([var1, var2])`?
- Not sure if returning the indices of the boolvars is very user-friendly but saving the variables in the solver also seems like a bad idea, are there any other good alternatives?

### CODE REVIEW

File: ortools/sat/csharp/CpModel.cs [hunk 1]
```diff
@@ -615,6 +615,11 @@ public void AddHint(IntVar var, long value)
             model_.SolutionHint.Values.Add(value);
         }
 
+        public void AddAssumption(IntVar var)
```
Comment (lperron) [inline ortools/sat/csharp/CpModel.cs:618]: Wrong type. Maybe we should accept a iLiteral, and make sure the literal is positive.

File: ortools/sat/python/cp_model.py [hunk 2]
```diff
@@ -1542,6 +1542,9 @@ def AddHint(self, var, value):
         self.__model.solution_hint.vars.append(self.GetOrMakeIndex(var))
```
Comment (lperron) [inline ortools/sat/python/cp_model.py:1542]: We need to make sure it is positive

File: ortools/sat/samples/assumptions_sample_sat.cc [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  const BoolVar b = cp_model.NewBoolVar().WithName("b");
+  // [END variables]
+
+  // [START constraints]
+  cp_model.AddGreaterThan(x, 10).OnlyEnforceIf(a);
+  cp_model.AddLessOrEqual(y, 10).OnlyEnforceIf(b);
+  // [END constraints]
+
+  // Add assumptions
+  cp_model.AddAssumption(a);
```
Comment (lperron) [inline ortools/sat/samples/assumptions_sample_sat.cc:40]: You are right, it should be nice to have a vector/list/enumerable API.

File: ortools/java/com/google/ortools/sat/CpModel.java [hunk 4]
```diff
@@ -966,6 +966,13 @@ public void addHint(IntVar var, long value) {
     modelBuilder.getSolutionHintBuilder().addValues(value);
   }
 
+  /** Adds variable as assumption */
+  public void addAssumptions(Literal[] literals) {
+    for (Literal lit : literals) {
+      modelBuilder.addAssumptions(lit.getIndex());
+    }
+  }
```
Comment (lperron) [inline ortools/java/com/google/ortools/sat/CpModel.java:974]: I would add both the single var, and the var[] API.

**Comment by stradivari96** [comment]
Thanks for the fast review! I have added the proposed changes

File: ortools/sat/cp_model.cc [hunk 5]
```diff
@@ -796,6 +796,12 @@ void CpModelBuilder::AddHint(IntVar var, int64 value) {
   cp_model_.mutable_solution_hint()->add_values(value);
 }
 
+void CpModelBuilder::AddAssumptions(absl::Span<const BoolVar> literals) {
+  for (const BoolVar& lit : literals) {
```
Comment (lperron) [inline ortools/sat/cp_model.cc:800]: Single var API and this one. Thanks

File: ortools/java/com/google/ortools/sat/CpModel.java (hunk 4 above)
Comment (stradivari96) [reply ortools/java/com/google/ortools/sat/CpModel.java:974]: Okey!

File: ortools/sat/cp_model.cc [hunk 6]
```diff
@@ -796,6 +796,12 @@ void CpModelBuilder::AddHint(IntVar var, int64 value) {
...
+    cp_model_.mutable_assumptions()->Add(lit.index_);
```
Comment (lperron) [inline ortools/sat/cp_model.cc:801]: ->Add(PositiveRef(lit.index_));

File: ortools/java/com/google/ortools/sat/CpModel.java (hunk 4 above)
Comment (stradivari96) [reply ortools/java/com/google/ortools/sat/CpModel.java:974]: Overload or two methods (without the s)?

File: ortools/sat/csharp/CpModel.cs [hunk 7]
```diff
@@ -615,6 +615,14 @@ public void AddHint(IntVar var, long value)
...
         }
 
+        public void AddAssumptions(IEnumerable<ILiteral> literals)
+        {
+            foreach (ILiteral lit in literals)
+            {
+                model_.Assumptions.Add(lit.GetIndex());
+            }
+        }
+
```
Comment (lperron) [inline ortools/sat/csharp/CpModel.cs:625]: add single var API, checks Positivity: 
if (lit.GetIndex() >= 0) {
  model_.Assumptions.Add(lit.GetIndex());
} else {
  model_.Assumptions.Add(lit.Not().GetIndex());
}

to be done in python, java, and .NET

**Comment by lperron** [comment]
Thanks for doing this.

**Comment by lperron** [comment]
2 methods, 1 with S, 1 without

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le jeu. 4 févr. 2021 à 20:30, Xiang Chen <notifications@github.com> a
écrit :

File: ortools/sat/csharp/CpModel.cs (hunk 7 above)
Comment (stradivari96) [reply ortools/sat/csharp/CpModel.cs:625]: Doesn't this change the semantics for a (nonsensical) case like this?
```python
x = model.NewIntVar(0, 10, "")
a = model.NewBoolVar("")

model.Add(x > 10).OnlyEnforceIf(a.Not())
model.Proto().assumptions.append(a.Not().Index())
```
How about raise an error instead?

File: ortools/sat/python/cp_model.py [hunk 8]
```diff
@@ -1544,8 +1544,17 @@ def AddHint(self, var, value):
...
+        self.__model.ClearField("solution_hint")
+
+    def AddAssumption(self, lit):
+        self.__model.assumptions.append(self.GetOrMakeBooleanIndex(lit))
+
+    def AddAssumptions(self, literals):
+        for lit in literals:
+            self.AddAssumption(lit)
+
+    def ClearAssumptions(self):
```
Comment (lperron) [inline ortools/sat/python/cp_model.py:1556]: I suppose we need this method in all languages.

**Comment by lperron** [comment]
Thanks for the effort.

two things: 
for the example, I would do the following:
3 vars (x, y, z), 3 literals (a, b, c). a => x > y, b => y > z, c => z > x. All 3 literals are assumptions.

**Comment by lperron** [comment]
second thing, returning the index raises a lot of question.  Keep it that way for now. I will revisit this.

**Comment by stradivari96** [comment]
Almost there! I think the only remaining thing is my question about Positive Refs.

**Comment by lperron** [comment]
Let me merge and tweak it further.

Thanks!!!

**Comment by stradivari96** [comment]
Thanks for your guidance!

**Comment by lperron** [comment]
It turns out the model accepts positive and negative literals as
assumptions.

I will push some minor fixes and start looking at reading the result from
the model.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le ven. 5 févr. 2021 à 17:03, Xiang Chen <notifications@github.com> a
écrit :

**Comment by stradivari96** [comment]
In case it helps, as an user I think something like `solver.ValidAssumption(lit) -> bool` should be good, not sure what to do if you return multiples unsat cores though.

**Pull request merged**